rasa/core/training/story_conflict.py

```diff
@@ -0,0 +1,235 @@
+
+
+class TrackerEventStateTuple(NamedTuple):
+    """Holds a tracker, an event, and sliced states associated with those."""
+
+    tracker: TrackerWithCachedStates
+    event: Event
+    sliced_states: List[Dict[Text, float]]
+
+    @property
+    def sliced_states_hash(self):
+        return hash(str(list(self.sliced_states)))
+
+
+def find_story_conflicts(
+    trackers: List[TrackerWithCachedStates], domain: Domain, max_history: int
+) -> List[StoryConflict]:
+    """Generates a list of StoryConflict objects, describing conflicts in the given trackers.
+
+    Args:
+        trackers: Trackers in which to search for conflicts.
+        domain: The domain.
+        max_history: The maximum history length to be taken into account.
+    Returns:
+        List of conflicts.
+    """
+
+    # We do this in two steps, to reduce memory consumption:
+
+    # Create a 'state -> list of actions' dict, where the state is
+    # represented by its hash
+    state_action_mapping = _find_conflicting_states(trackers, domain, max_history)
+
+    # Iterate once more over all states and note the (unhashed) state,
+    # for which a conflict occurs
+    conflicts = _build_conflicts_from_states(
+        trackers, domain, max_history, state_action_mapping
+    )
+
+    return conflicts
+
+
+def _find_conflicting_states(
+    trackers: List[TrackerWithCachedStates], domain: Domain, max_history: int
+) -> Dict[int, Optional[List[Text]]]:
+    # Create a 'state -> list of actions' dict, where the state is
+    # represented by its hash
+    state_action_mapping = defaultdict(list)
+    for element in _sliced_states_iterator(trackers, domain, max_history):
+        hashed_state = element.sliced_states_hash
+        if element.event.as_story_string() not in state_action_mapping[hashed_state]:
+            state_action_mapping[hashed_state] += [element.event.as_story_string()]
+
+    # Keep only conflicting `state_action_mapping`s
+    return {
+        state_hash: actions
+        for (state_hash, actions) in state_action_mapping.items()
+        if len(actions) > 1
+    }
+
+
+def _build_conflicts_from_states(
+    trackers: List[TrackerWithCachedStates],
+    domain: Domain,
+    max_history: int,
+    state_action_dict: Dict[int, Optional[List[Text]]],
+) -> List["StoryConflict"]:
+    # Iterate once more over all states and note the (unhashed) state,
+    # for which a conflict occurs
+    conflicts = {}
+    for element in _sliced_states_iterator(trackers, domain, max_history):
+        hashed_state = element.sliced_states_hash
+
+        if hashed_state in state_action_dict:
+            if hashed_state not in conflicts:
+                conflicts[hashed_state] = StoryConflict(element.sliced_states)
+
+            conflicts[hashed_state].add_conflicting_action(
+                action=element.event.as_story_string(),
+                story_name=element.tracker.sender_id,
+            )
+
+    # Remove conflicts that arise from unpredictable actions
```

Comment:
optional: the comment is a bit misleading since you're also changing the type from a dictionary to a list